src/types/list.rs

```diff
@@ -720,4 +773,124 @@ mod tests {
             assert!(list.index(42i32).is_err());
         });
     }
+
+    use std::ops::Range;
+
+    struct FaultyIter(Range<usize>, usize);
+
+    impl Iterator for FaultyIter {
+        type Item = usize;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            self.0.next()
+        }
+    }
+
+    impl ExactSizeIterator for FaultyIter {
+        fn len(&self) -> usize {
+            self.1
+        }
+    }
+
+    #[test]
+    #[should_panic(
+        expected = "Attempted to create PyList but `elements` was larger than reported by its `ExactSizeIterator` implementation."
+    )]
+    fn too_long_iterator() {
+        Python::with_gil(|py| {
+            let iter = FaultyIter(0..usize::MAX, 73);
+            let _list = PyList::new(py, iter);
+        })
+    }
+
+    #[test]
+    #[should_panic(
+        expected = "Attempted to create PyList but `elements` was smaller than reported by its `ExactSizeIterator` implementation."
+    )]
+    fn too_short_iterator() {
+        Python::with_gil(|py| {
+            let iter = FaultyIter(0..35, 73);
+            let _list = PyList::new(py, iter);
+        })
+    }
+
+    #[test]
+    #[should_panic(
+        expected = "out of range integral type conversion attempted on `elements.len()`"
+    )]
+    fn overflowing_size() {
+        Python::with_gil(|py| {
+            let iter = FaultyIter(0..0, usize::MAX);
+
+            let _list = PyList::new(py, iter);
+        })
+    }
+
+    #[cfg(feature = "macros")]
+    #[test]
+    fn bad_clone_mem_leaks() {
```

Comment:
So do I gather correctly from this test that decrementing the list pointer even though it's not fully initialized is fine? (The NULL / uninitialized members don't cause segfault?)

Reply:
> So do I gather correctly from this test that decrementing the list pointer even though it's not fully initialized is fine?

The [documentation](https://docs.python.org/3/c-api/list.html#c.PyList_New) explicitly mentions its items are `NULL` when initially initialized, rather than unitialized.

> The NULL / uninitialized members don't cause segfault?

`list_dealloc` uses `Py_XDECREF` to drop its items (see https://github.com/python/cpython/blob/main/Objects/listobject.c#L352), so a `PyList` having `NULL` items shouldn't be a problem. However I don't see this guaranteed anywhere.

Reply:
Yep makes sense, I think it's fine for us to assume it's sound then.